File: src/calc/aggregate.ts
import type { Accum, Breakdown, Factor, SeqFactor } from "../types/Sequence";

const accumulate = (factor: Factor, prevValue: number, value: number) => {
  if (factor === "plus") {
    return prevValue + value;
  } else if (factor === "minus") {
    return prevValue - value;
  } else {
    throw new Error();
  }
};

const addBreakdown = (breakdown: Breakdown, factor: SeqFactor) => {
  if (factor.tag in breakdown) {
    breakdown[factor.tag] = {
      ids: [...breakdown[factor.tag].ids, factor.id],
      store: accumulate(
        factor.factor,
        breakdown[factor.tag].store,
        factor.value,
      )
    };
  } else {
    breakdown[factor.tag] = { ids: [factor.id], store: accumulate(factor.factor, 0, factor.value) };
  }
  return breakdown;
};

export const aggregate = (
  sequence: SeqFactor[],
  baseValue: number,
  filter: string[],
): Accum[] => {
  if (sequence.length < 1) return [];

  const seq = sequence;
  const accums: Accum[] = [];
  const sorted = seq.sort((a, b) => a.time.getTime() - b.time.getTime());

  const [firstFactor, ...restFactors] = sorted;

  let timePos = firstFactor.time;
  let accum: Accum = {
    ids: [firstFactor.id],
    tags: [firstFactor.tag],
    time: timePos,
    store: accumulate(firstFactor.factor, baseValue, firstFactor.value),
    breakdown: {
      [firstFactor.tag]: { ids: [firstFactor.id], store: accumulate(firstFactor.factor, 0, firstFactor.value) },
    },
  };

  for (const seqFactor of restFactors) {
    if (filter.length > 0 && !filter.includes(seqFactor.tag)) continue;

    const factor = seqFactor.factor;
    const value = seqFactor.value;
    const prevValue = accum.store;

    if (timePos === seqFactor.time) {
      accum = {
        ids: [...accum.ids, seqFactor.id],
        tags: Array.from(new Set([...accum.tags, seqFactor.tag])),
        time: timePos,
        store: accumulate(factor, prevValue, value),
        breakdown: addBreakdown(accum.breakdown, seqFactor),
      };
    } else {
      accums.push(accum);
      const newTimePos = seqFactor.time;
      accum = {
        ids: [seqFactor.id],
        tags: [seqFactor.tag],
        time: newTimePos,
        store: accumulate(factor, prevValue, value),
        breakdown: {
          [seqFactor.tag]: { ids: [seqFactor.id], store: accumulate(seqFactor.factor, 0, seqFactor.value) },
        },
      };
      timePos = newTimePos;
    }
  }
  if (accum !== null) {
    accums.push(accum);
  }
  return accums;
};

File: aggregate.go
package aggrett

// Aggregate groups factors by exact time and returns running totals with per-time breakdown.
// If filter is empty, all tags are included. Otherwise only listed tags are included.
func Aggregate(sequence []SeqFactor, baseValue float64, filter []string) ([]Accum, error) {
	return aggregate(sequence, baseValue, filter, nil)
}

// AggregateByInterval groups factors into interval buckets and returns running totals with per-bucket breakdown.
// Result times are the start of each bucket.
func AggregateByInterval(sequence []SeqFactor, baseValue float64, filter []string, step int, intervalType IntervalType) ([]Accum, error) {
	return aggregate(sequence, baseValue, filter, &timeGrouping{
		Step:         step,
		IntervalType: intervalType,
	})
}

func aggregate(sequence []SeqFactor, baseValue float64, filter []string, grouping *timeGrouping) ([]Accum, error) {
	filtered := sequence
	if len(filter) > 0 {
		filtered = FilterByTag(sequence, filter)
	}

	groups, err := groupByTime(filtered, grouping)
	if err != nil {
		return nil, err
	}
	if len(groups) == 0 {
		return []Accum{}, nil
	}

	result := make([]Accum, 0, len(groups))
	store := baseValue
	for _, group := range groups {
		ids := make([]string, 0, len(group.Factors))
		tags := make([]string, 0, len(group.Factors))
		tagsSeen := make(map[string]struct{})
		breakdown := make(Breakdown)

		for _, f := range group.Factors {
			ids = append(ids, f.ID)
			store, err = Accumulate(f.Factor, store, f.Value)
			if err != nil {
				return nil, err
			}

			if _, ok := tagsSeen[f.Tag]; !ok {
				tagsSeen[f.Tag] = struct{}{}
				tags = append(tags, f.Tag)
			}

			entry := breakdown[f.Tag]
			delta, err := Accumulate(f.Factor, entry.Delta, f.Value)
			if err != nil {
				return nil, err
			}
			breakdown[f.Tag] = BreakdownEntry{
				IDs:   append(entry.IDs, f.ID),
				Delta: delta,
			}
		}

		result = append(result, Accum{
			IDs:       ids,
			Tags:      tags,
			Time:      group.Time,
			Store:     store,
			Breakdown: breakdown,
		})
	}
	return result, nil
}
